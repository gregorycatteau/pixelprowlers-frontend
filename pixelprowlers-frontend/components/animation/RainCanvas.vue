<template>
  <canvas ref="canvas" class="rain-canvas"></canvas>
</template>

<script setup lang="ts">
// Import des hooks Vue
import { onMounted, ref, onBeforeUnmount } from 'vue'

/**
 * Référence au canvas
 */
const canvas = ref<HTMLCanvasElement | null>(null)

onMounted(() => {
  const ctx = canvas.value?.getContext('2d')
  if (!ctx || !canvas.value) return

  const fontSize = 24
  const drops: number[] = []

  // Nombre de colonnes basé sur la largeur
  const columns = Math.floor(window.innerWidth / fontSize)
  for (let i = 0; i < columns; i++) drops[i] = Math.random() * 40

  /**
   * Fonction pour adapter la taille du canvas lors du resize
   */
  const resizeCanvas = () => {
    if (canvas.value) {
      canvas.value.width = window.innerWidth
      canvas.value.height = window.innerHeight
    }
  }

  // Initialisation de la taille
  resizeCanvas()
  window.addEventListener('resize', resizeCanvas)

  /**
   * Fonction de dessin des gouttes
   */
  const draw = () => {
    // Nettoie sans remplir
    ctx.clearRect(0, 0, canvas.value!.width, canvas.value!.height)

    // Paramètres visuels des gouttes
    ctx.fillStyle = 'rgba(255, 255, 255, 0.15)'
    ctx.shadowColor = 'rgba(255, 255, 255, 0.2)'
    ctx.shadowBlur = 4
    ctx.font = `${fontSize}px monospace`

    for (let i = 0; i < drops.length; i++) {
      const text = '|'
      const x = i * fontSize
      const y = drops[i] * fontSize

      ctx.fillText(text, x, y)

      if (y > canvas.value!.height || Math.random() > 0.95) {
        drops[i] = 0
      }
      drops[i] += 1
    }
  }

  // Interval de rafraîchissement
  const interval = setInterval(draw, 33)

  onBeforeUnmount(() => {
    clearInterval(interval)
    window.removeEventListener('resize', resizeCanvas)
  })
})
</script>

<style scoped>
.rain-canvas {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 2;
  pointer-events: none;
  mix-blend-mode: screen; /* Ou 'lighten' si ton fond est sombre */
}
</style>
