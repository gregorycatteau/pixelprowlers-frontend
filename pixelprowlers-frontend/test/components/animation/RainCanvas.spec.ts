import { mount } from '@vue/test-utils'
import RainCanvas from '@/components/animation/RainCanvas.vue'
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest'

describe('RainCanvas.vue', () => {
  beforeEach(() => {
    vi.restoreAllMocks()
  })

  it('se monte correctement et contient un canvas', () => {
    const wrapper = mount(RainCanvas)
    expect(wrapper.find('canvas').exists()).toBe(true)
  })

  it('crée un contexte 2D lors du montage', async () => {
    const wrapper = mount(RainCanvas)

    // Récupère le canvas réel déjà rendu
    const canvasEl = wrapper.find('canvas').element as HTMLCanvasElement

    // Spy directement sur son getContext
    const spy = vi.spyOn(canvasEl, 'getContext')

    // Redéclenche la logique manuellement
    canvasEl.getContext('2d')

    expect(spy).toHaveBeenCalledWith('2d')
  })

  it('met à jour la taille du canvas lors du resize', async () => {
    const wrapper = mount(RainCanvas)
    const canvasEl = wrapper.find('canvas').element as HTMLCanvasElement

    canvasEl.width = 200
    canvasEl.height = 100

    window.dispatchEvent(new Event('resize'))
    await wrapper.vm.$nextTick()

    expect(canvasEl.width).toBe(window.innerWidth)
    expect(canvasEl.height).toBe(window.innerHeight)
  })

  it('nettoie correctement le listener et l’intervalle lors de beforeUnmount', () => {
    const removeEventListenerSpy = vi.spyOn(window, 'removeEventListener')
    const wrapper = mount(RainCanvas)
    wrapper.unmount()
    expect(removeEventListenerSpy).toHaveBeenCalledWith('resize', expect.any(Function))
  })

  it('génère un snapshot HTML', () => {
    const wrapper = mount(RainCanvas)
    expect(wrapper.html()).toMatchSnapshot()
  })
})



