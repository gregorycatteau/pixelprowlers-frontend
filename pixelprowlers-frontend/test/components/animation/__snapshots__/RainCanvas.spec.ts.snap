// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RainCanvas.vue > génère un snapshot HTML 1`] = `"<canvas data-v-5a7eb9f7="" class="rain-canvas" width="1024" height="768"></canvas>"`;
